README.md
# Addon connector for Kodi (script.module.addon.connector)

[![Kodi version](https://img.shields.io/badge/kodi%20versions-19-blue)](https://kodi.tv/)
[![GitHub release](https://img.shields.io/github/release/CastagnaIT/script.module.addon.connector.svg)](https://github.com/CastagnaIT/script.module.addon.connector/releases)
[![CI](https://github.com/CastagnaIT/script.module.addon.connector/workflows/CI/badge.svg)](https://github.com/CastagnaIT/script.module.addon.connector/actions?query=workflow:CI)
[![Codecov](https://img.shields.io/codecov/c/github/CastagnaIT/script.module.addon.connector/master)](https://codecov.io/gh/CastagnaIT/script.module.addon.connector/branch/master)
[![License: LGPL-2.1 or later](https://img.shields.io/badge/license-LGPLv2.1_or_later-blue)](https://opensource.org/licenses/LGPL-2.1)
[![Contributors](https://img.shields.io/github/contributors/CastagnaIT/script.module.addon.connector.svg)](https://github.com/CastagnaIT/script.module.addon.connector/graphs/contributors)

A Kodi module to provide communication between add-ons and add-ons services.

## Main features

- Allows exchanging data between an add-on to his service or another add-on service
- Allows exchanging data between an add-on to another add-on
- Allows exchanging data between multiple add-on services
- Allows call a function of another add-on or service (no return data)
- With the default data serialization ([pickle](https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled))
you can transfer the most python data types
- Supports automatic forwarding of the exceptions to the add-on that made the call

## How integrate it into your add-on

A quick example of add-on communication, to allow data to be sent and received between the add-on and its service.

We assume that you have already developed a Kodi add-on with [his service](https://kodi.wiki/view/Service_add-ons),
then in the add-on script, we make the call to his service in order to send and ask some data: 

```python
import addonconnector
import xbmc

CALL_CFG = addonconnector.CallConfig(None)

def ask_data_to_service():
    ret_data = addonconnector.make_call(CALL_CFG.signal_name('the_service_function'),
                                        'sun', color='yellow')
    xbmc.log('Response received from the service: ' + ret_data)
```

On the service script of the add-on, we handle the callback, to answer to the add-on call request:

```python
import addonconnector
import xbmc

def the_service_function(name, color='blue'):
    xbmc.log('The add-on has asked if the {} is {}.'.format(name, color))
    data = 'Yes it is!' if color == 'yellow' else 'Nope!'
    # Here you can return any kind of data also the raised exceptions will be forwarded
    return data

if __name__ == '__main__':
    monitor = xbmc.Monitor()
    addonconnector.register_callback(the_service_function)
    while not monitor.abortRequested():
    # ...follow the Kodi add-on service development example
```

You can find all the detailed instructions and others examples are on the Wiki pages.

## Download links

Install add-on via repository - provide automatic installation of updates:

Todo

## License

Licensed under GNU Lesser General Public License version 2.1.

## Credits

This module is based on the "Addon Signals" module, thanks to Rick Phillips (Ruuk) who developed Addon Signals module:<br/>
https://github.com/ruuk/script.module.addon.signals


changelog.txt
v0.0.6
Removed wrong 1e-6 multiplication (Thanks CastagnaIT)
Fix wrong timeout on linux arm devices (Thanks CastagnaIT)
Introduced optional WaitTimeoutError exception on timeout (Thanks CastagnaIT)
Avoid blocking Kodi when abort is requested (Thanks CastagnaIT)

v0.0.5
Fixes an issue with json.loads() on Python 3.5 (Thanks Dag Wieers)
Use executeJSONRPC() instead of executebuiltin() (Thanks Dag Wieers)
Dev: Add sanity and unit checks (Thanks Dag Wieers)
Dev: Add setup.py for automated testing (Thanks Dag Wieers)

v0.0.4
Python 3 compatibility (thanks to romanvm)

v0.0.3
Make add-on backwards compatible

v0.0.2
Add registerCall(), makeCall() and returnCall()

v0.0.1
Initial release


test/xbmc.py
# -*- coding: utf-8 -*-
# Copyright: (c) 2019, Dag Wieers (@dagwieers) <dag@wieers.com>
# GNU General Public License v3.0 (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
"""This file implements the Kodi xbmc module, either using stubs or alternative functionality"""

# pylint: disable=bad-option-value,too-few-public-methods,useless-object-inheritance
from __future__ import absolute_import, division, print_function, unicode_literals
import json
import time
import weakref

LOGLEVELS = ['Debug', 'Info', 'Notice', 'Warning', 'Error', 'Severe', 'Fatal', 'None']
LOGDEBUG = 0
LOGINFO = 1
LOGNOTICE = 2
LOGWARNING = 3
LOGERROR = 4
LOGSEVERE = 5
LOGFATAL = 6
LOGNONE = 7


class Monitor(object):
    """A stub implementation of the xbmc Monitor class"""
    _instances = set()

    def __init__(self, line='', heading=''):  # pylint: disable=unused-argument
        """A stub constructor for the xbmc Monitor class"""
        self._instances.add(weakref.ref(self))

    @classmethod
    def getinstances(cls):
        """Return the instances for this class"""
        dead = set()
        for ref in cls._instances:
            obj = ref()
            if obj is not None:
                yield obj
            else:
                dead.add(ref)
        cls._instances -= dead


def executeJSONRPC(jsonrpccommand):
    """A reimplementation of the xbmc executeJSONRPC() function"""
    command = json.loads(jsonrpccommand)

    ret = dict(id=command.get('id'), jsonrpc='2.0', result='OK')
    if command.get('method') == 'JSONRPC.NotifyAll':
        # Send a notification to all instances of subclasses
        for sub in Monitor.__subclasses__():
            for obj in sub.getinstances():
                obj.onNotification(
                    sender=command.get('params').get('sender'),
                    method=command.get('params').get('message'),
                    data=json.dumps(command.get('params').get('data')),
                )
    else:
        log("executeJSONRPC does not implement method '{method}'".format(**command), LOGERROR)
        return json.dumps(dict(error=dict(code=-1, message='Not implemented'), id=command.get('id'), jsonrpc='2.0'))
    return json.dumps(ret)


def log(msg, level=0):
    """A reimplementation of the xbmc log() function"""
    color1 = '\033[32;1m'
    color2 = '\033[32;0m'
    name = LOGLEVELS[level]
    if level in (4, 5, 6, 7):
        color1 = '\033[31;1m'
        if level in (6, 7):
            raise Exception(msg)
    elif level in (2, 3):
        color1 = '\033[33;1m'
    elif level == 0:
        color2 = '\033[30;1m'
    print('{color1}{name}: {color2}{msg}\033[39;0m'.format(name=name, color1=color1, color2=color2, msg=msg))


def sleep(timemillis):
    """A reimplementation of the xbmc sleep() function"""
    time.sleep(timemillis / 1000)


tests/xbmc.py
# -*- coding: utf-8 -*-
# Copyright: (c) 2019, Dag Wieers (@dagwieers) <dag@wieers.com>
# GNU General Public License v3.0 (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
"""This file implements the Kodi xbmc module, either using stubs or alternative functionality"""

# pylint: disable=bad-option-value,too-few-public-methods,useless-object-inheritance,invalid-name
import json
import time
import weakref

LOGLEVELS = ['Debug', 'Info', 'Notice', 'Warning', 'Error', 'Severe', 'Fatal', 'None']
LOGDEBUG = 0
LOGINFO = 1
LOGNOTICE = 2
LOGWARNING = 3
LOGERROR = 4
LOGSEVERE = 5
LOGFATAL = 6
LOGNONE = 7


class Monitor:
    """A stub implementation of the xbmc Monitor class"""
    _instances = set()

    def __init__(self, line='', heading=''):  # pylint: disable=unused-argument
        """A stub constructor for the xbmc Monitor class"""
        self.iteration = 0
        self._instances.add(weakref.ref(self))

    def abortRequested(self):
        """A stub implementation for the xbmc Keyboard class abortRequested() method"""
        self.iteration += 1
        # print('Iteration: %s' % self.iteration)
        # return self.iteration % 5 == 0
        return False

    def waitForAbort(self, timeout=None):  # pylint: disable=no-self-use
        """A stub implementation for the xbmc Monitor class waitForAbort() method"""
        try:
            time.sleep(timeout)
        except KeyboardInterrupt:
            return True
        except Exception:  # pylint: disable=broad-except
            return True
        return False

    @classmethod
    def getinstances(cls):
        """Return the instances for this class"""
        dead = set()
        for ref in cls._instances.copy():
            obj = ref()
            if obj is not None:
                yield obj
            else:
                dead.add(ref)
        cls._instances -= dead


def executeJSONRPC(jsonrpccommand):
    """A reimplementation of the xbmc executeJSONRPC() function"""
    command = json.loads(jsonrpccommand)

    ret = dict(id=command.get('id'), jsonrpc='2.0', result='OK')
    if command.get('method') == 'JSONRPC.NotifyAll':
        # Send a notification to all instances of subclasses
        for sub in Monitor.__subclasses__():
            for obj in sub.getinstances():
                obj.onNotification(
                    sender=command.get('params').get('sender'),
                    method='Other.' + command.get('params').get('message'),
                    data=json.dumps(command.get('params').get('data')),
                )
    else:
        log("executeJSONRPC does not implement method '{method}'".format(**command), LOGERROR)
        return json.dumps(dict(error=dict(code=-1, message='Not implemented'), id=command.get('id'), jsonrpc='2.0'))
    return json.dumps(ret)


def log(msg, level=0):
    """A reimplementation of the xbmc log() function"""
    color1 = '\033[32;1m'
    color2 = '\033[32;0m'
    name = LOGLEVELS[level]
    if level in (4, 5, 6, 7):
        color1 = '\033[31;1m'
        if level in (6, 7):
            raise Exception(msg)
    elif level in (2, 3):
        color1 = '\033[33;1m'
    elif level == 0:
        color2 = '\033[30;1m'
    print('{color1}{name}: {color2}{msg}\033[39;0m'.format(name=name, color1=color1, color2=color2, msg=msg))


def sleep(timemillis):
    """A reimplementation of the xbmc sleep() function"""
    time.sleep(timemillis / 1000)


test/xbmcaddon.py
# -*- coding: utf-8 -*-
# Copyright: (c) 2019, Dag Wieers (@dagwieers) <dag@wieers.com>
# GNU General Public License v3.0 (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
"""This file implements the Kodi xbmcaddon module, either using stubs or alternative functionality"""

# pylint: disable=invalid-name,too-few-public-methods

from __future__ import absolute_import, division, print_function, unicode_literals

ADDON_ID = 'plugin.video.foobar'


class Addon:
    """A reimplementation of the xbmcaddon Addon class"""

    def __init__(self, id=None):  # pylint: disable=redefined-builtin
        """A stub constructor for the xbmcaddon Addon class"""
        self.id = id or ADDON_ID

    def getAddonInfo(self, key):
        """A working implementation for the xbmcaddon Addon class getAddonInfo() method"""
        addon_info = dict(id=self.id, name='Foobar', version='1.2.3', type='kodi.video', profile='special://userdata', path='special://userdata')
        return addon_info.get(key)


lib/AddonSignals.py
# -*- coding: utf-8 -*-
# GNU Lesser General Public License v2.1 (see COPYING or https://www.gnu.org/licenses/gpl-2.1.txt)
"""The AddonSignals module provides signal/slot mechanism for inter-addon communication in Kodi"""

from xbmc import executeJSONRPC, log, LOGNOTICE, Monitor, sleep


class WaitTimeoutError(Exception):
    """ Exception used when waiting times out"""


def _perf_clock():
    """Provides high resolution timing in seconds"""
    import time
    if hasattr(time, 'perf_counter'):
        return time.perf_counter()  # pylint: disable=no-member
    if hasattr(time, 'clock'):
        # time.clock() was deprecated in Python 3.3 and removed in Python 3.8
        return time.clock()  # pylint: disable=no-member
    return time.time()  # Fallback


def _addon_id():
    """Return the Kodi add-on id and cache it as a static variable"""
    if not hasattr(_addon_id, 'cached'):
        from xbmcaddon import Addon
        _addon_id.cached = Addon().getAddonInfo('id')
    return getattr(_addon_id, 'cached')


def _to_unicode(text, encoding='utf-8', errors='strict'):
    """Force text to unicode"""
    if isinstance(text, bytes):
        return text.decode(encoding, errors=errors)
    return text


def _receiver():
    """Return a SignalReceiver instance and cache it as a static variable"""
    if not hasattr(_receiver, 'cached'):
        _receiver.cached = SignalReceiver()
    return getattr(_receiver, 'cached')


def _decode_data(data):
    """Decode base64-encoded JSON data and return Python data structure"""
    import json
    encoded_data = json.loads(data)
    if not encoded_data:
        return None
    from base64 import b64decode
    json_data = b64decode(encoded_data[0])
    # NOTE: With Python 3.5 and older json.loads() does not support bytes or bytearray
    return json.loads(_to_unicode(json_data))


def _encode_data(data):
    """Encode Python data structure into base64-encoded JSON data"""
    from base64 import b64encode
    import json
    json_data = json.dumps(data)
    if not isinstance(json_data, bytes):
        json_data = json_data.encode('utf-8')
    encoded_data = b64encode(json_data)
    return encoded_data.decode('ascii')


def _jsonrpc(**kwargs):
    """Perform JSONRPC calls"""
    import json
    if 'id' not in kwargs:
        kwargs.update(id=0)
    if 'jsonrpc' not in kwargs:
        kwargs.update(jsonrpc='2.0')
    return json.loads(executeJSONRPC(json.dumps(kwargs)))


class SignalReceiver(Monitor, object):  # pylint: disable=bad-option-value,useless-object-inheritance
    """The AddonSignals receiver class"""
    def __init__(self):  # pylint: disable=super-init-not-called
        self._slots = {}
        super(SignalReceiver, self).__init__()  # pylint: disable=bad-option-value,super-with-arguments

    def registerSlot(self, signaler_id, signal, callback):
        """Register a slot in the AddonSignals receiver"""
        if signaler_id not in self._slots:
            self._slots[signaler_id] = {}
        self._slots[signaler_id][signal] = callback

    def unRegisterSlot(self, signaler_id, signal):
        """Unregister a slot in the AddonSignals receiver"""
        if signaler_id not in self._slots:
            return
        if signal not in self._slots[signaler_id]:
            return
        del self._slots[signaler_id][signal]

    def onNotification(self, sender, method, data):
        """The Kodi Monitor event handler for notifications"""
        if not sender.endswith('.SIGNAL'):
            return
        sender = sender[:-7]
        if sender not in self._slots:
            return
        signal = method.split('.', 1)[-1]
        if signal not in self._slots[sender]:
            return
        self._slots[sender][signal](_decode_data(data))


class CallHandler(object):  # pylint: disable=bad-option-value,useless-object-inheritance
    """The AddonSignals event handler class"""
    def __init__(self, signal, data, source_id, timeout=1000, use_timeout_exception=False):  # pylint: disable=too-many-arguments
        """The CallHandler constructor"""
        self.signal = signal
        self.data = data
        self.timeout = timeout
        self.source_id = source_id
        self._return = None
        self.is_callback_received = False
        self.use_timeout_exception = use_timeout_exception
        registerSlot(self.source_id, '_return_{0}'.format(self.signal), self.callback)
        sendSignal(signal, data, self.source_id)

    def callback(self, data):
        """Method to register function as callback"""
        self._return = data
        self.is_callback_received = True

    def waitForReturn(self):
        """Wait for callback to trigger"""
        monitor = Monitor()
        end_time = _perf_clock() + (self.timeout / 1000)
        while not self.is_callback_received:
            if _perf_clock() > end_time:
                if self.use_timeout_exception:
                    unRegisterSlot(self.source_id, self.signal)
                    raise WaitTimeoutError
                break
            if monitor.abortRequested():
                raise OSError
            sleep(10)
        unRegisterSlot(self.source_id, self.signal)
        return self._return


def registerSlot(signaler_id, signal, callback):
    """
    Register a slot for a function callback
    :param signaler_id: the name used for call/answer (e.g. add-on id)
    :param signal: name of the function to call (can be the same used in returnCall/makeCall/...)
    :param callback: the function to call
    """
    _receiver().registerSlot(signaler_id, signal, callback)


def unRegisterSlot(signaler_id, signal):
    """API method to unregister a slot"""
    _receiver().unRegisterSlot(signaler_id, signal)


def sendSignal(signal, data=None, source_id=None, sourceID=None):
    """API method to send a signal"""
    if sourceID:
        log('++++==== script.module.addon.signals: sourceID keyword is DEPRECATED - use source_id ====++++', LOGNOTICE)
    _jsonrpc(method='JSONRPC.NotifyAll', params=dict(
        sender='%s.SIGNAL' % (source_id or sourceID or _addon_id()),
        message=signal,
        data=[_encode_data(data)],
    ))


def registerCall(signaler_id, signal, callback):
    """API method to register a callback slot"""
    registerSlot(signaler_id, signal, callback)


def returnCall(signal, data=None, source_id=None):
    """
    Make a return call to the target add-on
    :param signal: name of the function to call (can be the same used in registerSlot/makeCall/...)
    :param data: data to send
    :param source_id: the name used for call/answer (e.g. add-on id)
    """
    sendSignal('_return_{0}'.format(signal), data, source_id)


def makeCall(signal, data=None, source_id=None, timeout_ms=1000, use_timeout_exception=False):
    """
    Make a call to the source add-on
    :param signal: name of the function to call (can be the same used in registerSlot/returnCall/...)
    :param data: data to send
    :param source_id: the name used for call/answer (e.g. add-on id)
    :param timeout_ms: maximum waiting time before the timeout
    :param use_timeout_exception: if True when the timeout occurs will raise the exception 'TimeoutError'
             (allow to return 'None' value from the callback data)
    """
    return CallHandler(signal, data, source_id, timeout_ms, use_timeout_exception).waitForReturn()
